src/Chainweb/Pact/Backend/PactState/GrandHash.hs

```diff
@@ -0,0 +1,753 @@
+              $ Map.toList state
+        let hash = hashRows (Text.encodeUtf8 tblName) rows
+        pure hash
+
+  -- This is a simple incremental hash over all of the table hashes.
+  -- This is well-defined by its order because 'getLatestPactStateUpperBound''
+  -- guarantees that the stream of tables is ordered lexicographically by table
+  -- name.
+  S.fold_
+    (\ctx hash -> hashUpdate ctx (hashWith SHA3_256 hash))
+    (hashInitWith SHA3_256)
+    (Memory.convert . hashFinalize)
+    hashStream
+
+-- | Incremental hash over a vector
+hashAggregate :: (a -> ByteString) -> Vector a -> Maybe ByteString
+hashAggregate f v
+  | Vector.length v == 0 = Nothing
+  | otherwise = Just
+      $ Memory.convert
+      $ hashFinalize
+      $ Vector.foldl'
+          (\ctx a -> hashUpdate ctx (hashWith SHA3_256 (f a)))
+          (hashInitWith SHA3_256)
+          v
+
+-- | This is the grand hash of a table
+--
+--   Precondition: The PactRows must be in ordered by rowkey ASC
+hashRows :: ByteString -> Vector PactRow -> Maybe ByteString
+hashRows tblName = hashAggregate (rowToHashArgs tblName)
+
+-- | Turn a (TableName, PactRow) into the arguments for a hash function.
+--
+--   The format is:
+--
+--   char 'T' (1 byte)
+--   <length tablename> (8 bytes)
+--   <tablename> (variable, maximum TBD)
+--
+--   char 'K' (1 byte)
+--   <length rowkey> (8 bytes)
+--   <rowkey> (variable, maximum size TBD)
+--
+--   char 'I' (1 byte)
+--   <txId> (8 bytes)
+--
+--   char 'D' (1 byte)
+--   <rowdata> (variable, maximum size TBD)
+--
+--   And thus the size of this will always be 28 bytes + the total size of
+--   all variable sized inputs. None of the variable-sized inputs can
+--   be empty (other than maybe rowdata?).
+rowToHashArgs :: ByteString -> PactRow -> ByteString
+rowToHashArgs tblName pr =
+  BL.toStrict
+  $ BB.toLazyByteString
+  $ BB.charUtf8 'T'
+    <> BB.word64LE (fromIntegral @Int @Word64 (BS.length tblName))
+    <> BB.byteString tblName
+
+    <> BB.charUtf8 'K'
+    <> BB.word64LE (fromIntegral @Int @Word64 (BS.length pr.rowKey))
+    <> BB.byteString pr.rowKey
+
+    <> BB.charUtf8 'I'
+    <> BB.word64LE (fromIntegral @Int64 @Word64 pr.txId)
+
+    <> BB.charUtf8 'D'
+    <> BB.byteString pr.rowData
+
+-- | Get the BlockHeaders for each chain at the specified BlockHeight.
+getBlockHeadersAt :: ()
+  => RocksDb
+     -- ^ RocksDB handle
+  -> Map ChainId (Set BlockHeight)
+     -- ^ Map from ChainId to available blockheights on that chain.
+     --   This is most probably going to be the output of 'resolveTargets'.
+  -> BlockHeight
+     -- ^ The blockheight that you want to look up the header for.
+  -> ChainwebVersion
+     -- ^ chainweb version
+  -> IO (Map ChainId BlockHeader)
+getBlockHeadersAt rdb resolvedTargets bh v = do
+  wbhdb <- initWebBlockHeaderDb rdb v
+  let cutHashes = cutHashesTable rdb
+  -- Get the latest cut
+  highestCuts <- readHighestCutHeaders v (\_ _ -> pure ()) wbhdb cutHashes
```

Comment:
not sure if we want to actually pass a logging function here. answer is probably yes. it would be a trivial refactor, and may save time if this fails for some reason.